File: OpenEvolve-Plugin/src/nodes/KnowledgeQueryNode.ts
/**
 * Knowledge Query Node
 *
 * Knowledge graph query node for intelligent information retrieval.
 * Supports semantic search, entity extraction, and relationship analysis.
 *
 * @module nodes
 */

import {
  OpenEvolveBaseNode,
  NodeInputs,
  NodeResult,
  ExecutionContext,
  ValidationError,
  ParameterSchema
} from './OpenEvolveBaseNode';
import { knowledgeApi } from '@/services/api/endpoints';

/**
 * Query types for knowledge retrieval
 */
export type QueryType = 'semantic' | 'entity' | 'relationship' | 'path' | 'aggregation';

/**
 * Knowledge query node configuration
 */
export interface KnowledgeQueryNodeConfig {
  queryType?: QueryType;
  maxResults?: number;
  minRelevance?: number;
  enableReasoning?: boolean;
  includeMetadata?: boolean;
}

/**
 * Knowledge entity
 */
export interface KnowledgeEntity {
  id: string;
  type: string;
  label: string;
  properties: Record<string, any>;
  confidence: number;
}

/**
 * Knowledge relationship
 */
export interface KnowledgeRelationship {
  id: string;
  source: string;
  target: string;
  type: string;
  properties: Record<string, any>;
  confidence: number;
}

/**
 * Knowledge path result
 */
export interface KnowledgePath {
  entities: KnowledgeEntity[];
  relationships: KnowledgeRelationship[];
  length: number;
  confidence: number;
}

/**
 * Query result
 */
export interface KnowledgeQueryResult {
  queryId: string;
  queryType: QueryType;
  query: string;
  results: Array<{
    content: string;
    relevance: number;
    entities?: KnowledgeEntity[];
    source?: string;
    metadata?: Record<string, any>;
  }>;
  entities: KnowledgeEntity[];
  relationships: KnowledgeRelationship[];
  paths?: KnowledgePath[];
  summary: {
    totalResults: number;
    avgRelevance: number;
    entityCount: number;
    relationshipCount: number;
  };
  reasoning?: string;
  metadata: {
    executedAt: Date;
    executionTime: number;
    parameters: {
      maxResults: number;
      minRelevance: number;
      enableReasoning: boolean;
    };
  };
}

/**
 * Knowledge Query Node
 *
 * Queries the knowledge graph for relevant information.
 * Supports multiple query types and advanced reasoning.
 */
export class KnowledgeQueryNode extends OpenEvolveBaseNode {
  static readonly DISPLAY_NAME = 'Knowledge Query';
  static readonly DESCRIPTION = 'Query knowledge graph for semantic information, entities, and relationships';
  static readonly ICON = 'knowledge';
  static readonly CATEGORY = 'knowledge';
  static readonly VERSION = '1.0.0';

  constructor(id: string, config: KnowledgeQueryNodeConfig = {}) {
    super(id, {
      queryType: 'semantic',
      maxResults: 10,
      minRelevance: 0.5,
      enableReasoning: false,
      includeMetadata: true,
      ...config
    });
  }

  /**
   * Execute knowledge query
   *
   * @param inputs - Must contain 'query' string
   * @param context - Execution context
   * @returns Promise resolving to query result
   */
  async execute(inputs: NodeInputs, context: ExecutionContext): Promise<NodeResult> {
    try {
      const startTime = Date.now();

      // Extract inputs
      const query = inputs.query as string;
      const queryType = (inputs.queryType as QueryType) || (this.config.queryType as QueryType);
      const maxResults = (inputs.maxResults as number) || this.config.maxResults as number;
      const minRelevance = (inputs.minRelevance as number) || this.config.minRelevance as number;
      const enableReasoning = inputs.enableReasoning as boolean ?? this.config.enableReasoning as boolean;
      const filters = inputs.filters as Record<string, any> | undefined;

      // Validate required inputs
      if (!query || query.trim().length === 0) {
        return this.createErrorResult('Query is required and cannot be empty');
      }

      context.updateProgress(10, 'Preparing knowledge query');

      // Prepare query parameters
      const queryParams: Record<string, any> = {
        query,
        query_type: queryType,
        max_results: maxResults,
        min_relevance: minRelevance,
        enable_reasoning: enableReasoning,
        include_metadata: this.config.includeMetadata as boolean
      };

      if (filters) {
        queryParams.filters = filters;
      }

      context.updateProgress(30, 'Executing knowledge query');

      // Execute query via API
      let response: any;
      if (queryType === 'semantic') {
        // Use RAG search for semantic queries
        const ragResults = await knowledgeApi.searchRag({ query });
        
        // Filter and limit results
        const filteredResults = (ragResults || [])
          .filter((r: any) => {
            const score = r.score !== undefined ? r.score : (r.relevance !== undefined ? r.relevance : 0.5);
            return score >= minRelevance;
          })
          .slice(0, maxResults);
          
        response = { results: filteredResults };
      } else if (['relationship', 'path'].includes(queryType)) {
        // Use Graphiti for graph-based queries
        const graphResult = await knowledgeApi.searchGraphiti({ query, num_results: maxResults });
        
        if (graphResult && graphResult.nodes) {
          // Transform graph result to expected format
          response = {
            results: [],
            entities: (graphResult.nodes || []).map((n: any) => ({ label: n.id, ...n })),
            relationships: (graphResult.edges || []).map((e: any) => ({ 
               source: e.source, 
               target: e.target, 
               type: e.label, 
               confidence: 1.0 
            }))
          };
        } else {
          response = { results: [], entities: [], relationships: [] };
        }
      } else {
        // Fallback or generic query (implementation pending backend support)
        // For now, try RAG as a safe default
        const ragResults = await knowledgeApi.searchRag({ query });
        response = { results: ragResults };
      }

      context.updateProgress(70, 'Processing query results');

      // Process results
      const result = this.processQueryResult(response, query, queryType, startTime, enableReasoning);

      context.updateProgress(100, `Query complete: ${result.summary.totalResults} results found`);

      return this.createSuccessResult(result);

    } catch (error) {
      return this.createErrorResult(
        error instanceof Error ? error.message : 'Unknown error during knowledge query'
      );
    }
  }

  /**
   * Process query result from API
   *
   * @param response - API response
   * @param query - Original query
   * @param queryType - Query type
   * @param startTime - Execution start time
   * @param enableReasoning - Whether reasoning was enabled
   * @returns Processed query result
   */
  private processQueryResult(
    response: any,
    query: string,
    queryType: QueryType,
    startTime: number,
    enableReasoning: boolean
  ): KnowledgeQueryResult {
    const executionTime = Date.now() - startTime;
    const safeResponse = response || {};

    // Extract results from response
    const results = (safeResponse.results || []).map((r: any) => ({
      content: r.content || '',
      relevance: r.relevance !== undefined ? r.relevance : (r.score !== undefined ? r.score : 0.5),
      entities: r.entities || [],
      source: r.source,
      metadata: r.metadata
    }));

    // Extract entities and relationships
    const entities = (safeResponse.entities || []).map((e: any) => ({
      id: e.id || '',
      type: e.type || 'unknown',
      label: e.label || '',
      properties: e.properties || {},
      confidence: e.confidence || 0.5
    }));

    const relationships = (safeResponse.relationships || []).map((r: any) => ({
      id: r.id || '',
      source: r.source || '',
      target: r.target || '',
      type: r.type || 'related_to',
      properties: r.properties || {},
      confidence: r.confidence || 0.5
    }));

    // Extract paths if available
    const paths = safeResponse.paths?.map((p: any) => ({
      entities: p.entities || [],
      relationships: p.relationships || [],
      length: p.length || 0,
      confidence: p.confidence || 0.5
    }));

    // Calculate summary statistics
    const totalResults = results.length;
    const avgRelevance = totalResults > 0
      ? results.reduce((sum, r) => sum + r.relevance, 0) / totalResults
      : 0;

    const summary = {
      totalResults,
      avgRelevance,
      entityCount: entities.length,
      relationshipCount: relationships.length
    };

    return {
      queryId: response.query_id || `query-${Date.now()}`,
      queryType,
      query,
      results,
      entities,
      relationships,
      paths,
      summary,
      reasoning: enableReasoning ? response.reasoning : undefined,
      metadata: {
        executedAt: new Date(),
        executionTime,
        parameters: {
          maxResults: this.config.maxResults as number,
          minRelevance: this.config.minRelevance as number,
          enableReasoning
        }
      }
    };
  }

  /**
   * Validate input data
   *
   * @param inputs - Input data to validate
   * @returns Array of validation errors
   */
  validateInputs(inputs: NodeInputs): ValidationError[] {
    const errors: ValidationError[] = [];

    if (!inputs.query) {
      errors.push({
        field: 'query',
        message: 'Query is required',
        severity: 'error'
      });
    }

    if (inputs.query && typeof inputs.query !== 'string') {
      errors.push({
        field: 'query',
        message: 'Query must be a string',
        severity: 'error'
      });
    }

    if (inputs.query && inputs.query.length < 3) {
      errors.push({
        field: 'query',
        message: 'Query is too short (minimum 3 characters)',
        severity: 'warning'
      });
    }

    // Validate query type
    if (inputs.queryType && typeof inputs.queryType === 'string') {
      const validTypes = ['semantic', 'entity', 'relationship', 'path', 'aggregation'];
      if (!validTypes.includes(inputs.queryType)) {
        errors.push({
          field: 'queryType',
          message: `Query type must be one of: ${validTypes.join(', ')}`,
          severity: 'error'
        });
      }
    }

    // Validate max results
    if (inputs.maxResults && typeof inputs.maxResults === 'number') {
      if (inputs.maxResults < 1 || inputs.maxResults > 100) {
        errors.push({
          field: 'maxResults',
          message: 'Max results must be between 1 and 100',
          severity: 'error'
        });
      }
    }

    // Validate min relevance
    if (inputs.minRelevance && typeof inputs.minRelevance === 'number') {
      if (inputs.minRelevance < 0 || inputs.minRelevance > 1) {
        errors.push({
          field: 'minRelevance',
          message: 'Min relevance must be between 0 and 1',
          severity: 'error'
        });
      }
    }

    return errors;
  }

  /**
   * Get JSON Schema for configuration parameters
   *
   * @returns Parameter schema
   */
  getParameterSchema(): ParameterSchema {
    return {
      type: 'object',
      properties: {
        queryType: {
          type: 'string',
          description: 'Type of knowledge query to execute',
          enum: ['semantic', 'entity', 'relationship', 'path', 'aggregation'],
          default: 'semantic'
        },
        maxResults: {
          type: 'number',
          description: 'Maximum number of results to return',
          minimum: 1,
          maximum: 100,
          default: 10
        },
        minRelevance: {
          type: 'number',
          description: 'Minimum relevance threshold for results (0-1)',
          minimum: 0,
          maximum: 1,
          default: 0.5
        },
        enableReasoning: {
          type: 'boolean',
          description: 'Enable reasoning and explanation',
          default: false
        },
        includeMetadata: {
          type: 'boolean',
          description: 'Include metadata in results',
          default: true
        }
      },
      required: []
    };
  }

  /**
   * Add knowledge to the graph
   *
   * @param data - Knowledge data to add
   * @returns Promise resolving to add result
   */
  async addKnowledge(data: {
    content: string;
    entities?: Array<{ type: string; label: string; properties?: Record<string, any> }>;
    relationships?: Array<{ source: string; target: string; type: string; properties?: Record<string, any> }>;
    metadata?: Record<string, any>;
  }): Promise<NodeResult> {
    try {
      const response = await knowledgeApi.add(data);
      return this.createSuccessResult(response);
    } catch (error) {
      return this.createErrorResult(
        error instanceof Error ? error.message : 'Failed to add knowledge'
      );
    }
  }

  /**
   * Update knowledge in the graph
   *
   * @param knowledgeId - Knowledge ID to update
   * @param data - Updated knowledge data
   * @returns Promise resolving to update result
   */
  async updateKnowledge(knowledgeId: string, data: any): Promise<NodeResult> {
    // Backend support pending for update operations
    return this.createErrorResult('Update operation not supported by backend yet');
  }

  /**
   * Delete knowledge from the graph
   *
   * @param knowledgeId - Knowledge ID to delete
   * @returns Promise resolving to deletion result
   */
  async deleteKnowledge(knowledgeId: string): Promise<NodeResult> {
    // Backend support pending for delete operations
    return this.createErrorResult('Delete operation not supported by backend yet');
  }

  /**
   * List knowledge artifacts
   *
   * @param params - Optional query parameters
   * @returns Promise resolving to list of knowledge artifacts
   */
  async listKnowledge(params?: {
    type?: string;
    limit?: number;
    offset?: number;
  }): Promise<NodeResult> {
    try {
      const response = await knowledgeApi.list(params);
      return this.createSuccessResult(response);
    } catch (error) {
      return this.createErrorResult(
        error instanceof Error ? error.message : 'Failed to list knowledge'
      );
    }
  }

  /**
   * Get knowledge statistics
   *
   * @returns Promise resolving to knowledge graph statistics
   */
  async getStatistics(): Promise<NodeResult> {
    try {
      const response = await knowledgeApi.getStatistics();
      return this.createSuccessResult(response);
    } catch (error) {
      return this.createErrorResult(
        error instanceof Error ? error.message : 'Failed to get statistics'
      );
    }
  }
}

export default KnowledgeQueryNode;

File: bubblelabs_nodes/knowledge_query_node.py
"""
Knowledge Query Node for BubbleLabs Integration

Provides querying capabilities for the Unified Knowledge Graph:
- Query triples by subject, predicate, or object
- Find paths between entities
- Get entity neighbors
- Export knowledge as JSON
"""

from typing import Dict, Any, List, Optional
from datetime import datetime
from .base_node import BubbleLabsNode, NodeExecutionError


class KnowledgeQueryNode(BubbleLabsNode):
    """
    Query the unified knowledge graph for entities, relationships, and paths.

    Supports four query types:
    - triples: Query triples by subject, predicate, or object with confidence filtering
    - paths: Find paths between two entities with configurable max length
    - neighbors: Get neighborhood of an entity with configurable depth
    - export: Export the entire knowledge graph as JSON

    All queries support minimum confidence thresholds for quality control.
    """

    # Node metadata
    DISPLAY_NAME = "Knowledge Query"
    DESCRIPTION = "Query the unified knowledge graph for entities, relationships, and paths"
    ICON = "knowledge-query"
    CATEGORY = "knowledge"
    VERSION = "1.0.0"

    def __init__(self, config: Dict[str, Any] = None):
        super().__init__(config)

        # Import UnifiedKnowledgeGraph using safe import pattern
        UnifiedKnowledgeGraph = self.safe_import(
            'knowledge_engine.graph.unified_kg.UnifiedKnowledgeGraph',
            fallback_value=None,
            error_msg="UnifiedKnowledgeGraph not available for KnowledgeQueryNode"
        )

        # Also try alternative import paths
        if UnifiedKnowledgeGraph is None:
            UnifiedKnowledgeGraph = self.safe_import(
                'graph.unified_kg.UnifiedKnowledgeGraph',
                fallback_value=None,
                error_msg="UnifiedKnowledgeGraph not found in alternate path"
            )

        if UnifiedKnowledgeGraph is None:
            # Try one more common location
            unified_kg_module = self.safe_import(
                'unified_kg',
                fallback_value=None,
                error_msg="unified_kg module not available"
            )
            if unified_kg_module:
                UnifiedKnowledgeGraph = getattr(unified_kg_module, 'UnifiedKnowledgeGraph', None)

        self.UnifiedKnowledgeGraph = UnifiedKnowledgeGraph

        # Initialize knowledge graph instance if available
        self.kg_instance = None
        if UnifiedKnowledgeGraph:
            try:
                backend = self.config.get('backend', 'networkx')
                self.kg_instance = UnifiedKnowledgeGraph(backend=backend)
                self.logger.info(f"KnowledgeQueryNode initialized with {backend} backend")
            except Exception as e:
                self.logger.warning(f"Could not initialize UnifiedKnowledgeGraph: {e}")
                self.kg_instance = None

    def validate_inputs(self, inputs: Dict) -> List[str]:
        """
        Validate input parameters based on query type.

        Required fields vary by query_type:
        - triples: No required fields (optional filters: subject, predicate, object)
        - paths: Requires source_entity and target_entity
        - neighbors: Requires source_entity (the center entity)
        - export: No required fields
        """
        errors = []

        # Get query type from inputs or config
        query_type = inputs.get('query_type')
        if query_type is None:
            query_type = self.config.get('query_type')

        if query_type is None:
            errors.append("Missing required field: query_type (must be 'triples', 'paths', 'neighbors', or 'export')")
            return errors

        valid_query_types = ['triples', 'paths', 'neighbors', 'export']
        if query_type not in valid_query_types:
            errors.append(f"Invalid query_type: {query_type}. Must be one of: {', '.join(valid_query_types)}")
            return errors

        # Validate query-type specific requirements
        if query_type == 'paths':
            source = inputs.get('source_entity') or self.config.get('source_entity')
            target = inputs.get('target_entity') or self.config.get('target_entity')
            if not source:
                errors.append("Path queries require 'source_entity' (in inputs or config)")
            if not target:
                errors.append("Path queries require 'target_entity' (in inputs or config)")

        elif query_type == 'neighbors':
            source = inputs.get('source_entity') or self.config.get('source_entity')
            if not source:
                errors.append("Neighbor queries require 'source_entity' (in inputs or config)")

        # Validate numeric parameters if provided
        if 'min_confidence' in inputs:
            try:
                confidence = float(inputs['min_confidence'])
                if not (0.0 <= confidence <= 1.0):
                    errors.append("min_confidence must be between 0.0 and 1.0")
            except (ValueError, TypeError):
                errors.append("min_confidence must be a number")

        if 'max_path_length' in inputs:
            try:
                length = int(inputs['max_path_length'])
                if length < 1:
                    errors.append("max_path_length must be at least 1")
            except (ValueError, TypeError):
                errors.append("max_path_length must be an integer")

        if 'depth' in inputs:
            try:
                depth = int(inputs['depth'])
                if depth < 1:
                    errors.append("depth must be at least 1")
            except (ValueError, TypeError):
                errors.append("depth must be an integer")

        return errors

    def execute(self, inputs: Dict, context) -> Dict[str, Any]:
        """
        Execute the knowledge query based on query_type.

        Args:
            inputs: Query specification including query_type and parameters
            context: Workflow state for tracking progress and artifacts

        Returns:
            Dict containing query results:
                - query_type: The type of query executed
                - results: Query-specific results (triples, paths, neighbors, or export data)
                - metadata: Query execution metadata
                - total_count: Number of results returned

        Raises:
            NodeExecutionError: If query execution fails
        """
        # Get query parameters
        query_type = inputs.get('query_type', self.config.get('query_type'))

        # Merge inputs with config (inputs take precedence)
        subject = inputs.get('subject') or self.config.get('subject')
        predicate = inputs.get('predicate') or self.config.get('predicate')
        obj = inputs.get('object') or self.config.get('object')
        min_confidence = inputs.get('min_confidence', self.config.get('min_confidence', 0.0))
        source_entity = inputs.get('source_entity') or self.config.get('source_entity')
        target_entity = inputs.get('target_entity') or self.config.get('target_entity')
        max_path_length = inputs.get('max_path_length', self.config.get('max_path_length', 3))
        depth = inputs.get('depth', self.config.get('depth', 1))

        context.update_progress(10, f"Initializing {query_type} query")
        self.logger.info(f"Executing {query_type} query")

        try:
            # Get the knowledge graph instance to use
            kg = self._get_knowledge_graph(inputs)

            if kg is None:
                raise NodeExecutionError(
                    node_name=self.get_display_name(),
                    message="No knowledge graph available for querying",
                    details={'query_type': query_type}
                )

            context.update_progress(30, "Executing query")

            # Execute the appropriate query
            if query_type == 'triples':
                result = self._query_triples(kg, subject, predicate, obj, min_confidence)
            elif query_type == 'paths':
                result = self._query_paths(kg, source_entity, target_entity, max_path_length, min_confidence)
            elif query_type == 'neighbors':
                result = self._query_neighbors(kg, source_entity, depth, min_confidence)
            elif query_type == 'export':
                result = self._export_knowledge(kg, min_confidence)
            else:
                raise NodeExecutionError(
                    node_name=self.get_display_name(),
                    message=f"Unknown query type: {query_type}",
                    details={'valid_types': ['triples', 'paths', 'neighbors', 'export']}
                )

            context.update_progress(90, "Processing results")

            # Add execution metadata
            result['metadata'] = {
                'query_type': query_type,
                'executed_at': datetime.now().isoformat(),
                'execution_id': self.execution_id,
                'parameters': {
                    'subject': subject,
                    'predicate': predicate,
                    'object': obj,
                    'min_confidence': min_confidence,
                    'source_entity': source_entity,
                    'target_entity': target_entity,
                    'max_path_length': max_path_length,
                    'depth': depth
                }
            }

            context.update_progress(100, f"Query complete: {result.get('total_count', 0)} results")
            self.logger.info(f"Knowledge query completed: {result.get('total_count', 0)} results")

            return result

        except NodeExecutionError:
            raise
        except Exception as e:
            self.logger.error(f"Knowledge query failed: {str(e)}", exc_info=True)
            raise NodeExecutionError(
                node_name=self.get_display_name(),
                message=f"Knowledge query failed: {str(e)}",
                details={
                    'query_type': query_type,
                    'exception_type': type(e).__name__
                }
            ) from e

    def _get_knowledge_graph(self, inputs: Dict) -> Optional[Any]:
        """
        Get the knowledge graph instance to use.

        Priority:
        1. kg_instance from inputs
        2. self.kg_instance (initialized in __init__)
        3. None (if not available)
        """
        # Check if a knowledge graph instance was provided in inputs
        if 'kg_instance' in inputs:
            return inputs['kg_instance']

        # Use the instance created in __init__
        return self.kg_instance

    def _query_triples(
        self,
        kg: Any,
        subject: Optional[str],
        predicate: Optional[str],
        obj: Optional[str],
        min_confidence: float
    ) -> Dict[str, Any]:
        """Query triples by subject, predicate, or object."""
        triples = kg.get_triples(
            subject=subject,
            predicate=predicate,
            object=obj,
            min_confidence=min_confidence
        )

        # Convert triples to dictionaries
        triples_data = []
        for t in triples:
            triples_data.append({
                'subject': t.subject,
                'predicate': t.predicate,
                'object': t.object,
                'confidence': t.confidence,
                'source': t.source,
                'timestamp': t.timestamp.isoformat() if hasattr(t.timestamp, 'isoformat') else str(t.timestamp),
                'metadata': t.metadata
            })

        return {
            'query_type': 'triples',
            'results': triples_data,
            'total_count': len(triples_data),
            'filters': {
                'subject': subject,
                'predicate': predicate,
                'object': obj,
                'min_confidence': min_confidence
            }
        }

    def _query_paths(
        self,
        kg: Any,
        source: str,
        target: str,
        max_length: int,
        min_confidence: float
    ) -> Dict[str, Any]:
        """Find paths between two entities."""
        paths = kg.find_paths(source, target, max_length)

        # Filter paths by confidence if needed
        filtered_paths = []
        for path in paths:
            # Check if all edges in path meet confidence threshold
            if all(edge.get('confidence', 1.0) >= min_confidence for edge in path):
                filtered_paths.append(path)

        return {
            'query_type': 'paths',
            'source_entity': source,
            'target_entity': target,
            'results': filtered_paths,
            'total_count': len(filtered_paths),
            'path_lengths': [len(p) for p in filtered_paths],
            'filters': {
                'max_path_length': max_length,
                'min_confidence': min_confidence
            }
        }

    def _query_neighbors(
        self,
        kg: Any,
        entity: str,
        depth: int,
        min_confidence: float
    ) -> Dict[str, Any]:
        """Get neighborhood of an entity."""
        neighbors = kg.get_neighbors(entity, depth)

        # Filter edges by confidence if needed
        if min_confidence > 0.0 and 'edges' in neighbors:
            neighbors['edges'] = [
                e for e in neighbors['edges']
                if e.get('confidence', 1.0) >= min_confidence
            ]
            # Recalculate node count based on filtered edges
            connected_nodes = set()
            for edge in neighbors['edges']:
                connected_nodes.add(edge.get('from'))
                connected_nodes.add(edge.get('to'))
            neighbors['node_count'] = len(connected_nodes)
            neighbors['edge_count'] = len(neighbors['edges'])

        return {
            'query_type': 'neighbors',
            'center_entity': entity,
            'results': neighbors,
            'total_count': neighbors.get('node_count', 0) - 1,  # Exclude center
            'filters': {
                'depth': depth,
                'min_confidence': min_confidence
            }
        }

    def _export_knowledge(
        self,
        kg: Any,
        min_confidence: float
    ) -> Dict[str, Any]:
        """Export knowledge graph as JSON."""
        export_data = kg.export_to_dict()

        # Filter by confidence if needed
        if min_confidence > 0.0 and 'triples' in export_data:
            original_count = len(export_data['triples'])
            export_data['triples'] = [
                t for t in export_data['triples']
                if t.get('confidence', 1.0) >= min_confidence
            ]
            export_data['filtered_count'] = len(export_data['triples'])
            export_data['filtered_out'] = original_count - len(export_data['triples'])

            # Update statistics
            if 'statistics' in export_data:
                export_data['statistics']['filtered_triple_count'] = len(export_data['triples'])

        return {
            'query_type': 'export',
            'results': export_data,
            'total_count': len(export_data.get('triples', [])),
            'filters': {
                'min_confidence': min_confidence
            }
        }

    def get_parameter_schema(self) -> Dict[str, Any]:
        """
        Get JSON schema for node parameters.

        Returns JSON schema for UI configuration including:
        - query_type: Type of query to execute
        - subject/predicate/object: Filters for triple queries
        - min_confidence: Minimum confidence threshold
        - source_entity/target_entity: For path queries
        - max_path_length: Maximum path length for path queries
        - depth: Neighborhood depth for neighbor queries
        """
        return {
            "type": "object",
            "title": "Knowledge Query Configuration",
            "description": "Configure knowledge graph query parameters",
            "properties": {
                "query_type": {
                    "type": "string",
                    "title": "Query Type",
                    "description": "Type of query to execute against the knowledge graph",
                    "enum": ["triples", "paths", "neighbors", "export"],
                    "enumNames": [
                        "Query Triples",
                        "Find Paths",
                        "Get Neighbors",
                        "Export Knowledge"
                    ],
                    "default": "triples"
                },
                "subject": {
                    "type": "string",
                    "title": "Subject",
                    "description": "Filter triples by subject entity (optional)",
                    "default": ""
                },
                "predicate": {
                    "type": "string",
                    "title": "Predicate",
                    "description": "Filter triples by predicate/relationship (optional)",
                    "default": ""
                },
                "object": {
                    "type": "string",
                    "title": "Object",
                    "description": "Filter triples by object entity (optional)",
                    "default": ""
                },
                "min_confidence": {
                    "type": "number",
                    "title": "Minimum Confidence",
                    "description": "Minimum confidence threshold for results (0.0-1.0)",
                    "minimum": 0.0,
                    "maximum": 1.0,
                    "default": 0.0
                },
                "source_entity": {
                    "type": "string",
                    "title": "Source Entity",
                    "description": "Starting entity for path or neighbor queries",
                    "default": ""
                },
                "target_entity": {
                    "type": "string",
                    "title": "Target Entity",
                    "description": "Target entity for path queries",
                    "default": ""
                },
                "max_path_length": {
                    "type": "number",
                    "title": "Maximum Path Length",
                    "description": "Maximum path length for path queries",
                    "minimum": 1,
                    "maximum": 10,
                    "default": 3
                },
                "depth": {
                    "type": "number",
                    "title": "Neighborhood Depth",
                    "description": "Depth of neighborhood for neighbor queries",
                    "minimum": 1,
                    "maximum": 5,
                    "default": 1
                },
                "backend": {
                    "type": "string",
                    "title": "Storage Backend",
                    "description": "Backend storage for knowledge graph",
                    "enum": ["networkx", "memory"],
                    "enumNames": [
                        "NetworkX (recommended)",
                        "In-Memory"
                    ],
                    "default": "networkx"
                }
            },
            "required": ["query_type"],
            "dependencies": {
                "query_type": {
                    "oneOf": [
                        {
                            "properties": {
                                "query_type": {"enum": ["triples"]}
                            },
                            "description": "Query triples by subject, predicate, or object"
                        },
                        {
                            "properties": {
                                "query_type": {"enum": ["paths"]}
                            },
                            "required": ["source_entity", "target_entity"],
                            "description": "Find paths between two entities"
                        },
                        {
                            "properties": {
                                "query_type": {"enum": ["neighbors"]}
                            },
                            "required": ["source_entity"],
                            "description": "Get neighborhood of an entity"
                        },
                        {
                            "properties": {
                                "query_type": {"enum": ["export"]}
                            },
                            "description": "Export entire knowledge graph as JSON"
                        }
                    ]
                }
            }
        }

    def is_healthy(self) -> bool:
        """
        Check if the node is healthy and ready to execute.

        Returns:
            True if UnifiedKnowledgeGraph is available, False otherwise
        """
        return self.UnifiedKnowledgeGraph is not None
